### ISSUE TITLE
`no_adjacent_strings_in_list` does not catch concatenation as part of addition

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#3738]
**Describe the issue**

The lint rule does not report an adjacent String which is one side of an addition of Strings.

**To Reproduce**

```dart
var list = <String>[
  'a' // should be LINT?
  'b' + 'c',
  'd' + 'e' // should be LINT?
  'f',
];
```

**Expected behavior**

I expect the adjacent Strings `'a' 'b'` and `'e' 'f'` to be reported.

I could certainly be wrong here. Why just `+`? It doesn't seem like arbitrarily deep adjacent Strings should be reported (like `[print('a' 'b')]` should be ok? The wrapping tokens make it obvious?). But what about `'a' 'b'.length`? With newlines, it seems like `['a', 'b' 'c'.length, 'd']` would be just as error-prone (if you're _not_ using the formatter) as direct adjacent String children of a List...

If we want to head down this road _at all_, maybe we should just add a new lint rule, `avoid_adjacent_strings`, which would also support https://github.com/dart-lang/language/issues/1982.

**Comment by bwilkerson** [comment]
The intent of the rule is to detect cases where the lack of a comma might indicate that two values that were intended to be separate elements of the list are being concatenated into a single element. Deeper adjacent strings can't be intended to be separate elements because neither the left nor the right would be a valid expression in that case.

Personally, I think that combining adjacent strings and `String.+` in the same concatenation is a code smell and might want to be a separate lint that applies everywhere, not just inside list and set literals.

But we probably want to wait a bit to see how the language issue is resolved.

**Comment by pq** [comment]
I agree with Brian.  This seems like a smell but probably a different lint.

**Comment by lrhn** [comment]
> The intent of the rule is to detect cases where the lack of a comma might indicate that two values that were intended to be separate elements of the list are being concatenated into a single element. 

I think that intent matches this case precisely. Maybe if it's *rephrased* as:

> The intent of the rule is to detect cases where the lack of a comma might indicate that two values that were intended to be *part of* separate elements of the list are being concatenated into a single element. 

If you write:
```dart
foo(
  "a" + "b"
  "c" + "d",
);
```
where you use `+` *because* of the lint, and you still forgot the comma after `b`, it's exactly the kind of concatenation the lint was intended to prevent.

The fact that the entire expression is not adjacent string literals is not important, it's that there are implicit concatenations in an expression in a list, which would also be valid if there was a comma between those two strings.

Harder to detect, certainly, but definitely within the intent of the lint.
